exp/addons/api/v1alpha3/clusterresourceset_webhook.go

```diff
@@ -0,0 +1,136 @@
+// +kubebuilder:webhook:verbs=create;update,path=/mutate-addons-cluster-x-k8s-io-v1alpha3-clusterresourceset,mutating=true,failurePolicy=fail,matchPolicy=Equivalent,groups=addons.cluster.x-k8s.io,resources=clusterresourcesets,versions=v1alpha3,name=default.clusterresourceset.addons.cluster.x-k8s.io,sideEffects=None
+
+var _ webhook.Defaulter = &ClusterResourceSet{}
+var _ webhook.Validator = &ClusterResourceSet{}
+
+// Default implements webhook.Defaulter so a webhook will be registered for the type
+func (m *ClusterResourceSet) Default() {
+	// ClusterResourceSet Strategy defaults to ApplyOnce.
+	if m.Spec.Strategy == "" {
+		m.Spec.Strategy = string(ClusterResourceSetStrategyApplyOnce)
+	}
+}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (m *ClusterResourceSet) ValidateCreate() error {
+	return m.validate(nil)
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
+func (m *ClusterResourceSet) ValidateUpdate(old runtime.Object) error {
+	oldCRS, ok := old.(*ClusterResourceSet)
+	if !ok {
+		return apierrors.NewBadRequest(fmt.Sprintf("expected a ClusterResourceSet but got a %T", old))
+	}
+	return m.validate(oldCRS)
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type
+func (m *ClusterResourceSet) ValidateDelete() error {
+	return m.validate(nil)
+}
+
+func (m *ClusterResourceSet) validate(old *ClusterResourceSet) error {
+	var allErrs field.ErrorList
+
+	// Validate selector parses as Selector
+	selector, err := metav1.LabelSelectorAsSelector(&m.Spec.ClusterSelector)
+	if err != nil {
+		allErrs = append(
+			allErrs,
+			field.Invalid(field.NewPath("spec", "clusterSelector"), m.Spec.ClusterSelector, err.Error()),
+		)
+	}
+
+	// Validate that the selector isn't empty.
+	if selector != nil && selector.Empty() {
+		allErrs = append(
+			allErrs,
+			field.Invalid(field.NewPath("spec", "selector"), m.Spec.ClusterSelector, "selector must not be empty"),
+		)
+	}
+
+	// Validate that the selector isn't empty as null selectors do not select any objects.
+	if selector != nil && selector.Empty() {
+		allErrs = append(
+			allErrs,
+			field.Invalid(field.NewPath("spec", "selector"), m.Spec.ClusterSelector, "selector must not be empty"),
+		)
+	}
+
+	// Only allow supported strategies
+	if m.Spec.Strategy != "" && !isSupportedStrategy(m.Spec.Strategy) {
+		allErrs = append(
+			allErrs,
+			field.NotSupported(field.NewPath("spec", "strategy"), m.Spec.Resources, []string{string(ClusterResourceSetStrategyApplyOnce)}),
+		)
+	}
+
+	if old != nil && old.Spec.Strategy != m.Spec.Strategy {
+		allErrs = append(
+			allErrs,
+			field.Invalid(field.NewPath("spec", "strategy"), m.Spec.Strategy, "field is immutable"),
+		)
+	}
+
+	// TODO: Any changes to clusterSelector is not allowed ?
```

Comment:
Do you mean we shouldn't allow it?

Reply:
Yes, I was not sure if we should allow any changes. Removed the comment.